usr/src/lib/libzfs/common/libzfs_sendrecv.c

```diff
@@ -1288,28 +1322,34 @@ dump_snapshot(zfs_handle_t *zhp, void *arg)
 
 	if (!sdd->dryrun) {
 		/*
-		 * If progress reporting is requested, spawn a new thread to
-		 * poll ZFS_IOC_SEND_PROGRESS at a regular interval.
+		 * Spawn a new thread to poll ZFS_IOC_SEND_PROGRESS at a regula
+		 * interval, or upon receiving SIGINFO signal.
 		 */
-		if (sdd->progress) {
-			pa.pa_zhp = zhp;
-			pa.pa_fd = sdd->outfd;
-			pa.pa_parsable = sdd->parsable;
+		pa.pa_zhp = zhp;
+		pa.pa_fd = sdd->outfd;
+		pa.pa_parsable = sdd->parsable;
+		pa.pa_progress = sdd->progress;
+		pa.pa_siginfo = sdd->siginfo;
 
-			if ((err = pthread_create(&tid, NULL,
+		if (pa.pa_siginfo) {
+			/* Make sure this thread ignores signal, or ioctl aborts */
+			sigset_t sigset;
+			(void) sigemptyset(&sigset);
+			(void) sigaddset(&sigset, SIGINFO);
+			(void) pthread_sigmask(SIG_BLOCK, &sigset, NULL);
+		}
+
+		if ((err = pthread_create(&tid, NULL,
 			    send_progress_thread, &pa)) != 0) {
```

Comment:
What if neither `progress` or `siginfo` is set?  Does the thread (correctly) do nothing then?

Also, looks like the indentation is wrong, continuation should be +4 spaces. (or maybe github is just rendering wrong)